File: csharp/ScreenplayTools.Tests/Fountain/ParserTest.cs
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

namespace ScreenplayTools.Tests;
using System.IO;
using ScreenplayTools;
using ScreenplayTools.Fountain;

public class ParserTest
{
    private string loadTestFile(string fileName) {
        return File.ReadAllText("../../../../../tests/"+fileName);
    }

    [Fact]
    public void Scratch()
    {
        string source = loadTestFile("Scratch.fountain");
        string match = loadTestFile("Scratch.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void SceneHeading()
    {
        string source = loadTestFile("SceneHeading.fountain");
        string match = loadTestFile("SceneHeading.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void ActionMerged()
    {
        string source = loadTestFile("Action.fountain");
        string match = loadTestFile("Action-Merged.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void ActionUnmerged()
    {
        string source = loadTestFile("Action.fountain");
        string match = loadTestFile("Action-Unmerged.txt");

        Parser fp = new Parser();
        fp.MergeActions = false;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Character()
    {
        string source = loadTestFile("Character.fountain");
        string match = loadTestFile("Character.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void DialogueMerged()
    {
        string source = loadTestFile("Dialogue.fountain");
        string match = loadTestFile("Dialogue-Merged.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void DialogueUnmerged()
    {
        string source = loadTestFile("Dialogue.fountain");
        string match = loadTestFile("Dialogue-Unmerged.txt");

        Parser fp = new Parser();
        fp.MergeDialogue = false;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Parenthetical()
    {
        string source = loadTestFile("Parenthetical.fountain");
        string match = loadTestFile("Parenthetical.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Lyrics()
    {
        string source = loadTestFile("Lyrics.fountain");
        string match = loadTestFile("Lyrics.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Transition()
    {
        string source = loadTestFile("Transition.fountain");
        string match = loadTestFile("Transition.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void TitlePage()
    {
        string source = loadTestFile("TitlePage.fountain");
        string match = loadTestFile("TitlePage.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void PageBreak()
    {
        string source = loadTestFile("PageBreak.fountain");
        string match = loadTestFile("PageBreak.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void LineBreaks()
    {
        string source = loadTestFile("LineBreaks.fountain");
        string match = loadTestFile("LineBreaks.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Notes()
    {
        string source = loadTestFile("Notes.fountain");
        string match = loadTestFile("Notes.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Boneyards()
    {
        string source = loadTestFile("Boneyards.fountain");
        string match = loadTestFile("Boneyards.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void Sections()
    {
        string source = loadTestFile("Sections.fountain");
        string match = loadTestFile("Sections.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void UTF8()
    {
        string source = loadTestFile("UTF8.fountain");
        string match = loadTestFile("UTF8.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void Tags()
    {
        string source = loadTestFile("Tags.fountain");
        string match = loadTestFile("Tags.txt");

        Parser fp = new Parser();
        fp.UseTags = true;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }
}

File: javascript/test/fountain/parser.test.js
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

import { loadTestFile } from '../../test/testUtils.js';
import { strict as assert } from 'assert';
import { FountainParser } from "../../src/fountain/parser.js";

describe('FountainParser', () => {

  describe('Scratch', () => {
    it('should match', () => {

      const source = loadTestFile('Scratch.fountain');
      const match = loadTestFile('Scratch.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('SceneHeading', () => {
    it('should match', () => {

      const source = loadTestFile('SceneHeading.fountain');
      const match = loadTestFile('SceneHeading.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Actions', () => {
    it('should match merged', () => {

      const source = loadTestFile('Action.fountain');
      const match = loadTestFile('Action-Merged.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });

    it('should match unmerged', () => {

      const source = loadTestFile('Action.fountain');
      const match = loadTestFile('Action-Unmerged.txt');

      let fp = new FountainParser();
      fp.mergeActions = false;
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });

  });

  describe('Character', () => {
    it('should match', () => {

      const source = loadTestFile('Character.fountain');
      const match = loadTestFile('Character.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Dialogue', () => {
    it('should match merged', () => {

      const source = loadTestFile('Dialogue.fountain');
      const match = loadTestFile('Dialogue-Merged.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });

    it('should match unmerged', () => {

      const source = loadTestFile('Dialogue.fountain');
      const match = loadTestFile('Dialogue-Unmerged.txt');

      let fp = new FountainParser();
      fp.mergeDialogue = false;
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Parenthetical', () => {
    it('should match', () => {

      const source = loadTestFile('Parenthetical.fountain');
      const match = loadTestFile('Parenthetical.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Lyrics', () => {
    it('should match', () => {

      const source = loadTestFile('Lyrics.fountain');
      const match = loadTestFile('Lyrics.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Transition', () => {
    it('should match', () => {

      const source = loadTestFile('Transition.fountain');
      const match = loadTestFile('Transition.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('TitlePage', () => {
    it('should match', () => {

      const source = loadTestFile('TitlePage.fountain');
      const match = loadTestFile('TitlePage.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('PageBreak', () => {
    it('should match', () => {

      const source = loadTestFile('PageBreak.fountain');
      const match = loadTestFile('PageBreak.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('LineBreaks', () => {
    it('should match', () => {

      const source = loadTestFile('LineBreaks.fountain');
      const match = loadTestFile('LineBreaks.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Notes', () => {
    it('should match', () => {

      const source = loadTestFile('Notes.fountain');
      const match = loadTestFile('Notes.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Boneyards', () => {
    it('should match', () => {

      const source = loadTestFile('Boneyards.fountain');
      const match = loadTestFile('Boneyards.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Sections', () => {
    it('should match', () => {

      const source = loadTestFile('Sections.fountain');
      const match = loadTestFile('Sections.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('UTF8', () => {
    it('should match', () => {

      const source = loadTestFile('UTF8.fountain');
      const match = loadTestFile('UTF8.txt');

      let fp = new FountainParser();
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

  describe('Tags', () => {
    it('should match', () => {

      const source = loadTestFile('Tags.fountain');
      const match = loadTestFile('Tags.txt');

      let fp = new FountainParser();
      fp.useTags = true;
      fp.addText(source);

      let output = fp.script.dump();
      //console.log(output);
      assert.equal(output, match);
    });
  });

});